File: UNIT-01/orchestrator/workspace.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"unit01/clients"
)

// Workspace tracks the active working directory and session state.
type Workspace struct {
	Path         string
	SessionID    string
	ProjectMap   string
	Instructions string
	Identity     string
	Active       bool
}

func NewWorkspace() *Workspace {
	return &Workspace{}
}

// Set activates a workspace at the given path.
func (w *Workspace) Set(path string) (string, error) {
	client := clients.NewSandboxClient()
	sessionID, err := client.SetWorkspace(path)
	if err != nil {
		w.Path = path
		w.SessionID = "local"
		w.Active = true
		return "local", nil
	}
	w.Path = path
	w.SessionID = sessionID
	w.Active = true

	indexer := clients.NewIndexerClient()
	if m, err := indexer.GetProjectMap(path); err == nil {
		w.ProjectMap = m
	}

	if data, err := os.ReadFile(filepath.Join(path, "UNIT-01.md")); err == nil {
		w.Instructions = string(data)
	}

	identity := ""
	if data, err := os.ReadFile(filepath.Join(path, "go.mod")); err == nil {
		identity += "--- GO.MOD ---\n" + string(data) + "\n"
	}
	if data, err := os.ReadFile(filepath.Join(path, "README.md")); err == nil {
		content := string(data)
		if len(content) > 1000 {
			content = content[:1000] + "... (truncated)"
		}
		identity += "--- README.MD ---\n" + content + "\n"
	}
	w.Identity = identity

	return sessionID, nil
}

func (w *Workspace) Refresh() {
	if !w.Active {
		return
	}
	indexer := clients.NewIndexerClient()
	m, err := indexer.GetProjectMap(w.Path)
	if err != nil {
		w.ProjectMap = fmt.Sprintf("[Indexer Error: failed to fetch project map: %v]", err)
	} else if m == "" {
		w.ProjectMap = "[Indexer Error: project map returned empty string]"
	} else {
		w.ProjectMap = m
	}
}

File: UNIT-01/orchestrator/src/workspace.rs
use crate::clients::indexer::IndexerClient;
use std::path::Path;

pub struct Workspace {
    pub path: String,
    pub session_id: String,
    pub project_map: String,
    pub instructions: String,
    pub identity: String,
    pub active: bool,
}

impl Workspace {
    pub fn new() -> Self {
        Self {
            path: String::new(),
            session_id: String::new(),
            project_map: String::new(),
            instructions: String::new(),
            identity: String::new(),
            active: false,
        }
    }

    pub async fn set(&mut self, path: &str) -> String {
        self.path = path.to_string();
        self.session_id = "local".to_string();
        self.active = true;

        let indexer = IndexerClient::new();
        if let Ok(m) = indexer.get_project_map(path).await {
            self.project_map = m;
        }

        let instr_path = Path::new(path).join("UNIT-01.md");
        if let Ok(data) = std::fs::read_to_string(&instr_path) {
            self.instructions = data;
        }

        self.identity = String::new();
        let go_mod_path = Path::new(path).join("go.mod");
        if let Ok(data) = std::fs::read_to_string(&go_mod_path) {
            self.identity.push_str("--- GO.MOD ---\n");
            self.identity.push_str(&data);
            self.identity.push('\n');
        }

        self.session_id.clone()
    }

    pub async fn refresh(&mut self) {
        if !self.active {
            return;
        }
        let indexer = IndexerClient::new();
        match indexer.get_project_map(&self.path).await {
            Ok(m) => self.project_map = m,
            Err(e) => self.project_map = format!("[Indexer Error: {}]", e),
        }
    }
}
